### PR TITLE
SNOW-1825584: Cache deps to speed up the builds

### PR DESCRIPTION
**Pull request by sfc-gh-jszczerbinski**
In this PR:
- We cache deps built on master
- We use cached deps on pull requests to speed up build times

### CODE REVIEW

File: .github/workflows/build-test.yml [hunk 1]
```diff
@@ -21,11 +21,25 @@ jobs:
                 cloud_provider: [ 'AWS', 'AZURE', 'GCP' ]
         steps:
             - uses: actions/checkout@v1
+            - name: Restore cached deps
+              id: cache-restore-deps
+              uses: actions/cache/restore@v3
```
Comment (sfc-gh-dprzybysz) [inline .github/workflows/build-test.yml:26]: there is version 4 available (also for save action) https://github.com/actions/cache

File: .github/workflows/build-test.yml [hunk 2]
```diff
@@ -81,13 +110,27 @@ jobs:
...
+              with:
+                path: dep-cache
+                key: ${{ matrix.build_type }}-${{ matrix.platform }}-${{ matrix.vs_version }}-${{ github.event.pull_request.base.sha }}-Win-dep-cache
+              if: github.event_name == 'pull_request'
```
Comment (sfc-gh-dprzybysz) [inline .github/workflows/build-test.yml:119]: should we always restore or should we check explicitly if the entry is available?

File: .github/workflows/build-test.yml [hunk 3]
```diff
@@ -114,11 +157,25 @@ jobs:
...
               env:
                 BUILD_TYPE: ${{ matrix.build_type }}
               run: ./ci/build_mac.sh
+            - name: Cache deps
+              id: cache-save-deps
+              uses: actions/cache/save@v3
+              with:
+                path: dep-cache
+                key: ${{ matrix.build_type }}-${{ github.sha }}-Mac-dep-cache
+              if: github.ref_name == github.event.repository.default_branch && matrix.cloud_provider == 'AWS'
```
Comment (sfc-gh-dprzybysz) [inline .github/workflows/build-test.yml:178]: we save to different place that we restore from - is it correct?

Comment (sfc-gh-jszczerbinski) [reply .github/workflows/build-test.yml:178]: We build and cache deps on master (commit sha = github.sha, but we restore on pull requests on master. On pull requests we will restore deps based on a base_ref - some commit on master (commit sha = github.event.pull_request.base.sha). So we restore and save from the same cache -> just not on the same workflow.

File: .github/workflows/build-test.yml (hunk 1 above)
Comment (sfc-gh-jszczerbinski) [reply .github/workflows/build-test.yml:26]: I had issues with version 4 while testing. Maybe we can upgrade later on.

**Review by sfc-gh-dprzybysz — APPROVED** [review]

**Review by sfc-gh-mkubik — APPROVED** [review]

**Pull request merged**